tauri/src/app.rs

```diff
@@ -21,12 +24,171 @@ pub use webview::{
 };
 pub use webview_manager::{WebviewDispatcher, WebviewManager};
 
-type InvokeHandler<A> = dyn Fn(WebviewManager<A>, String, JsonValue) -> BoxFuture<'static, crate::Result<InvokeResponse>>
-  + Send
-  + Sync;
-type ManagerHook<A> = dyn Fn(WebviewManager<A>) -> BoxFuture<'static, ()> + Send + Sync;
-type PageLoadHook<A> =
-  dyn Fn(WebviewManager<A>, PageLoadPayload) -> BoxFuture<'static, ()> + Send + Sync;
+type InvokeHandler<A> = dyn Fn(WebviewManager<A>, InvokeMessage<A>) + Send;
+type ManagerHook<A> = dyn Fn(WebviewManager<A>) + Send;
+type PageLoadHook<A> = dyn Fn(WebviewManager<A>, PageLoadPayload) + Send;
+
+/// Payload from an invoke call.
+#[derive(Debug, Deserialize)]
+pub(crate) struct InvokePayload {
+  #[serde(rename = "__tauriModule")]
+  tauri_module: Option<String>,
+  callback: String,
+  error: String,
+  #[serde(rename = "mainThread", default)]
+  pub(crate) main_thread: bool,
+  #[serde(flatten)]
+  inner: JsonValue,
+}
+
+/// An invoke message.
+pub struct InvokeMessage<A: ApplicationExt> {
```

Comment:
Is the reason for this type being public because of macro? If so, maybe we can add a doc(hidden) to it.

Reply:
this is a wrapper around the wry rpc and also provides APIs to resolve/reject the promise. I think I'll let it be documented since we do not force users to use the macros.